File: 0821-Shortest Distance to a Character/0821-Shortest Distance to a Character.py
class Solution:
    def shortestToChar(self, S, C):
        """
        :type S: str
        :type C: str
        :rtype: List[int]
        """
        n = len(S)
        result = []
        prev = -n
        for i, c in enumerate(S):
            if c == C:
                prev = i
            result.append(i - prev)
        
        prev = n * 2
        for i in range(n - 1, -1, -1):
            if S[i] == C:
                prev = i
            result[i] = min(result[i], prev - i)
        return result

File: 0801-0900/0821-Shortest Distance to a Character/0821-Shortest Distance to a Character.java
class Solution {
    public int[] shortestToChar(String S, char C) {
        int n = S.length();
        int[] result = new int[n];
        int prev = -n;
        for (int i = 0; i < n; ++i) {
            if (S.charAt(i) == C) {
                prev = i;
            }
            else {
                result[i] = i - prev;
            } 
        }
        
        prev = n * 2;
        for (int i = n - 1; i >= 0; --i) {
            if (S.charAt(i) == C) {
                prev = i;
            }
            else {
                result[i] = Math.min(result[i], prev - i);
            }   
        }
        return result;    
    }
}
